subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -217,17 +222,20 @@ private static void addHardSuccessorTasksToQueue(Node node, Set<Node> seenNodes,
     }
 
     private void collectValidationProblem(Node producer, Node consumer, TypeValidationContext validationContext, String consumerProducerPath) {
-        Severity severity = Severity.WARNING;
-        validationContext.visitPropertyProblem(
-            severity,
-            String.format("Gradle detected a problem with the following location: '%s'. "
-                    + "Task '%s' uses this output of task '%s' without declaring an explicit dependency (using Task.dependsOn() or Task.mustRunAfter()) or an implicit dependency (declaring task '%s' as an input). "
-                    + "This can lead to incorrect results being produced, depending on what order the tasks are executed",
-                consumerProducerPath,
-                consumer,
-                producer,
-                producer
-            )
+        validationContext.visitPropertyProblem(problem ->
+            problem.withId(ValidationProblemId.IMPLICIT_DEPENDENCY)
+                .reportAs(Severity.WARNING)
+                .withDescription(() -> "Gradle detected a problem with the following location: '" + consumerProducerPath + "'")
+                .happensBecause(() -> String.format("Task '%s' uses this output of task '%s' without declaring an explicit or implicit dependency or an implicit dependency. "
+                        + "This can lead to incorrect results being produced, depending on what order the tasks are executed",
+                    consumer,
+                    producer,
+                    producer
+                ))
+                .addPossibleSolution(() -> "Declare task '" + producer + "' as an input of '" + consumer + "'")
+                .addPossibleSolution(() -> "Declare an explicit dependency on '" + producer + "' to '" + consumer + "' using Task#dependsOn")
```

Comment:
```suggestion
                .addPossibleSolution(() -> "Declare an explicit dependency on '" + producer + "' from '" + consumer + "' using Task#dependsOn")
```

`from` sounds better to me than `to`. Same in the next solution.